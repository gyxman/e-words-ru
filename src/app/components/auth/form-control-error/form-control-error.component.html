<!-- Подпишемся на изменения, чтобы Angular обновлял представление,
иначе компонент не перерисуется из-за страгии OnPush и той же ссылки на объект FormControl -->
<ng-container *ngIf="control.statusChanges | async">
    <ng-container *ngIf="control.errors">
        <mat-error
            *ngIf="control.errors.required"
            automation-id="error-required"
            class="control-error"
        >
            Поле обязательно для заполнения
        </mat-error>
        <mat-error
            *ngIf="control.errors.email"
            automation-id="error-email"
            class="control-error"
        >
            Неверный формат E-mail
        </mat-error>
        <mat-error
            *ngIf="control.errors.minlength"
            automation-id="error-minlength"
            class="control-error"
        >
            Пароль должен быть не менее
            {{ control.errors.minlength.requiredLength }} символов. Сейчас
            {{ control.errors.minlength.actualLength }} символов
        </mat-error>
    </ng-container>
</ng-container>
