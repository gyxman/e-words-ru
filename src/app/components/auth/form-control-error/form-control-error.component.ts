import {ChangeDetectionStrategy, Component, Input} from '@angular/core';
import {FormControl} from '@angular/forms';

@Component({
    selector: 'app-form-control-error',
    templateUrl: './form-control-error.component.html',
    styleUrls: ['./form-control-error.component.less'],
    // https://github.com/angular/angular/issues/10887
    // changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FormControlErrorComponent {
    @Input() control: FormControl;
}
