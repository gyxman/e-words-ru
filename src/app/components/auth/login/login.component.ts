import {ChangeDetectionStrategy, Component} from '@angular/core';
import {FormControl, FormGroup, Validators} from '@angular/forms';

@Component({
    selector: 'app-login',
    templateUrl: './login.component.html',
    styleUrls: ['./login.component.less'],
    changeDetection: ChangeDetectionStrategy.OnPush,
})
export class LoginComponent {
    form = new FormGroup({
        email: new FormControl(null, [Validators.email, Validators.required]),
        password: new FormControl(null, [Validators.minLength(8), Validators.required]),
    });

    get emailControl(): FormControl {
        return this.form.get('email') as FormControl;
    }

    get passwordControl(): FormControl {
        return this.form.get('password') as FormControl;
    }
}
