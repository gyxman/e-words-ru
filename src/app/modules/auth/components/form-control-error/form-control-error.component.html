<ng-container *ngIf="control.touched && control.errors">
    <mat-error
        *ngIf="control.errors.required"
        automation-id="error-required"
        class="error"
    >
        Поле обязательно для заполнения
    </mat-error>
    <mat-error *ngIf="control.errors.email" automation-id="error-email" class="error">
        Неверный формат E-mail
    </mat-error>
    <mat-error
        *ngIf="control.errors.minlength"
        automation-id="error-minlength"
        class="error"
    >
        Пароль должен быть не менее
        {{ control.errors.minlength.requiredLength }} символов. Сейчас
        {{ control.errors.minlength.actualLength }} символов
    </mat-error>
</ng-container>
