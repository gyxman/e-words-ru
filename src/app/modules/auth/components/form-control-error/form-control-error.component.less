@import '../../styles/index';

:host {
    display: block;
}

.error {
    .font(13px, 20px);
}
