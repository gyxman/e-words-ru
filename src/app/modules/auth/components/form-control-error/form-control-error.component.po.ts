import {DebugElement} from '@angular/core';
import {PageObject} from '../../../../testing/page-object';

export class FormControlErrorComponentPo<T> extends PageObject<T> {
    get errorRequired(): DebugElement {
        return this.getByAutomationId('error-required');
    }

    get errorEmail(): DebugElement {
        return this.getByAutomationId('error-email');
    }

    get errorMinlength(): DebugElement {
        return this.getByAutomationId('error-minlength');
    }
}
