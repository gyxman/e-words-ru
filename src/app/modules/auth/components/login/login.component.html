<form automation-id="form" class="form" [formGroup]="form" (ngSubmit)="onSubmit()">
    <h2 class="form-title">
        Войти в тренажер
    </h2>
    <div class="form-content">
        <div class="form-row">
            <mat-form-field class="form-field">
                <mat-label>E-mail</mat-label>
                <input
                    automation-id="email-input"
                    matInput
                    placeholder="Введите e-mail"
                    [formControl]="emailControl"
                />
                <app-form-control-error
                    class="form-error"
                    [control]="emailControl"
                ></app-form-control-error>
            </mat-form-field>
        </div>
        <div class="form-row">
            <mat-form-field class="form-field">
                <mat-label>Пароль</mat-label>
                <input
                    automation-id="password-input"
                    type="password"
                    matInput
                    placeholder="Введите пароль"
                    [formControl]="passwordControl"
                />
                <app-form-control-error
                    class="form-error"
                    [control]="passwordControl"
                ></app-form-control-error>
            </mat-form-field>
        </div>
        <div class="form-row form-row_actions">
            <button
                automation-id="submit-button"
                mat-raised-button
                color="primary"
                class="button"
                type="submit"
                [disabled]="form.invalid"
            >
                Войти
            </button>
        </div>
    </div>
</form>
