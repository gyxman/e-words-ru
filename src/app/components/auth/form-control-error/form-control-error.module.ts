import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {FormControlErrorComponent} from './form-control-error.component';
import {MatFormFieldModule} from '@angular/material/form-field';

@NgModule({
    imports: [CommonModule, MatFormFieldModule],
    declarations: [FormControlErrorComponent],
    exports: [FormControlErrorComponent],
})
export class FormControlErrorModule {}
