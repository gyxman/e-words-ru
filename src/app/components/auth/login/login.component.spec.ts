import {ComponentFixture, TestBed} from '@angular/core/testing';

import {LoginComponent} from './login.component';
import {MockModule} from 'ng-mocks';
import {ReactiveFormsModule} from '@angular/forms';
import {MatInputModule} from '@angular/material/input';
import {MatButtonModule} from '@angular/material/button';
import {FormControlErrorModule} from '../form-control-error/form-control-error.module';

describe('LoginComponent - компонент входа в приложение', () => {
    let component: LoginComponent;
    let fixture: ComponentFixture<LoginComponent>;

    beforeEach(() => {
        TestBed.configureTestingModule({
            declarations: [LoginComponent],
            imports: [
                MockModule(ReactiveFormsModule),
                MockModule(MatInputModule),
                MockModule(MatButtonModule),
                MockModule(FormControlErrorModule),
            ],
        }).compileComponents();
    });

    beforeEach(() => {
        fixture = TestBed.createComponent(LoginComponent);
        component = fixture.componentInstance;
    });

    it('Если компонент загрузился, то все элементы отрисовались', () => {
        // act
        fixture.detectChanges();

        // assert
        expect(fixture.nativeElement).toMatchSnapshot();
    });
});
