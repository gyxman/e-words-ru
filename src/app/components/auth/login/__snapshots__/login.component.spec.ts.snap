// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent - компонент входа в приложение Если компонент загрузился, то все элементы отрисовались 1`] = `
<div
  id="root0"
>
  <form
    class="form"
  >
    <h2
      class="form-title"
    >
       Войти в тренажер 
    </h2>
    <div
      class="form-content"
    >
      <div
        class="form-row"
      >
        <mat-form-field
          class="form-field"
        >
          <mat-label>
            E-mail
          </mat-label>
          <input
            matinput=""
            placeholder="Введите e-mail"
          />
          <app-form-control-error
            class="form-error"
          />
        </mat-form-field>
      </div>
      <div
        class="form-row"
      >
        <mat-form-field
          class="form-field"
        >
          <mat-label>
            Пароль
          </mat-label>
          <input
            matinput=""
            placeholder="Введите пароль"
            type="password"
          />
          <app-form-control-error
            class="form-error"
          />
        </mat-form-field>
      </div>
      <div
        class="form-row form-row_actions"
      >
        <button
          class="button"
          color="primary"
          mat-raised-button=""
        >
           Войти 
        </button>
      </div>
    </div>
  </form>
</div>
`;
